modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj

```diff
@@ -174,8 +182,10 @@
     nil))
 
 (defmethod temporal-type Field
-  [{base-type :base_type, effective-type :effective_type, database-type :database_type}]
+  [{base-type :base_type, effective-type :effective_type, database-type :database_type,
+    coercion-strategy :coercion_strategy}]
   (or (database-type->temporal-type database-type)
+      (coercion-strategy->temporat-type coercion-strategy)
```

Comment:
I'm surprised this is necessary. When setting a corecion_strategy the effective_type should be set to the resulting type. (eg, base type is string, coercion is Coercion/ISO8601->DateTime, and effective is :type/DateTime or equivalent). The fact that this needed to be done makes me think we have some bug elsewhere if this is not happening.

Reply:
> The fact that this needed to be done makes me think we have some bug elsewhere if this is not happening.

You're right, and the bug is the mapping between our type system and the actual type we cast is mismatched for bigquery.

Specifically when you do a `:Coercion/ISO8601->DateTime` we do 2 things:
- we set the effective type of this column to `type/DateTime`
- we cast this column to `timestamp` ([ref](https://github.com/metabase/metabase/blob/a99ec1dfbb5bccaed4c827054df738789a0a0ea1/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L728))

The mismatch here is we interpret `type/Datetime` as `datetime` type, but the actually db-type is `timestamp` (because we cast to timestamp).

This is a problem now because bigquery distinguishes between `timestamp` and `datetime` when you do datetime arithmetic.
See [this](https://github.com/metabase/metabase/blob/8fa5b3c2e09aa9531058ff940e1ee019f63ea527/modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk/query_processor.clj#L721) comment for details. 

--- 

I think the correct fix here is to make sure the effective type for `:Coercion/ISO8601->DateTime` is `type/DateTimeWithTZ`, this way we will interpret it correctly as `timestamp`.

But the [code](https://github.com/metabase/metabase/blob/8fa5b3c2e09aa9531058ff940e1ee019f63ea527/shared/src/metabase/types.cljc#L355) for this is not dispatched per driver, And considering how a big `radius` this change affect our codebase, I think it's better to solve it within bigquery like I did here.

WDYT?

Reply:
actually, we could just make `:Coercion/ISO8601->DateTime` cast to `datetime` for bigquery.
Any reason this is a no-go?

Reply:
I went ahead with making `:Coercion/ISO8601->DateTime` to cast to `datetime` in this commit https://github.com/metabase/metabase/pull/25414/commits/786970c24f11c2ea8fe5277e42b9fa80b23a798e